lib/Dialect/HW/HWOps.cpp

```diff
@@ -1243,21 +1243,24 @@ SmallVector<Attribute> HWModuleOp::getAllPortAttrs() {
   auto attrs = getPerPortAttrs();
   if (attrs && !attrs->empty())
     return {attrs->getValue().begin(), attrs->getValue().end()};
-  return SmallVector<Attribute>(getNumPorts());
+  return SmallVector<Attribute>(getNumPorts(),
+                                DictionaryAttr::get(getContext()));
 }
 
 SmallVector<Attribute> HWModuleExternOp::getAllPortAttrs() {
   auto attrs = getPerPortAttrs();
   if (attrs && !attrs->empty())
     return {attrs->getValue().begin(), attrs->getValue().end()};
-  return SmallVector<Attribute>(getNumPorts());
+  return SmallVector<Attribute>(getNumPorts(),
+                                DictionaryAttr::get(getContext()));
```

Comment:
Invoking hash lookups and possible allocation is bad.  The getOrCreate cost of the empty dictionary isn't even amortized over the life of the module, it's per-call.

Reply:
I have little to no knowledge of the attribute infrastructure guts, but on a cursory look it appeared to me that retrieving the empty dictionary attribute should have little overhead. It is quite possible that I'm overlooking something here, but it seems to be effectively a member variable of the context and should not involve any hashing, as the call to the getter is immediately delegated to `DictionaryAttr::getEmpty`.